### PR TITLE
Simple drag-n-drop support on macOS and Windows

### PR DESCRIPTION
**Pull request by prime31**
Draft with new API per #345. This is tested working on macos and needs to be tested on Windows next.

### CODE REVIEW

**Comment by floooh** [comment]
Just a reminder for myself (I can easily fix this during merge):

The calls to calloc() and free() should be replaced with SOKOL_CALLOC and SOKOL_FREE macros (one day I actually want to get rid of SOKOL_CALLOC, and only have SOKOL_MALLOC, but that's for another time).

File: sokol_app.h [hunk 1]
```diff
@@ -4848,6 +4916,34 @@ _SOKOL_PRIVATE LRESULT CALLBACK _sapp_win32_wndproc(HWND hWnd, UINT uMsg, WPARAM
...
+                for (i = 0;  i < _sapp.num_dropped_files;  i++) {
+                    const UINT length = DragQueryFileW(drop, i, NULL, 0) + 1;
+                    WCHAR* buffer = (WCHAR*)SOKOL_CALLOC(length, sizeof(WCHAR));
+
+                    DragQueryFileW(drop, i, buffer, length);
+
+                    const int dst_needed = WideCharToMultiByte(CP_UTF8, 0, buffer, -1, 0, 0, NULL, NULL);
+                    int len = dst_needed > _sapp.max_dropped_file_path_length ? _sapp.max_dropped_file_path_length : dst_needed;
+                    _sapp_win32_wide_to_utf8(buffer, &_sapp.dropped_files[i * _sapp.max_dropped_file_path_length], len);
+                    free(buffer);
```
Comment (floooh) [inline sokol_app.h:4937]: ...this is the free() call that slipped through...

**Comment by floooh** [comment]
Ok I have started merging in branch https://github.com/floooh/sokol/tree/prime31-file-drop, apart from resolving conflicts (already done) I'll also do some changes related to the intenal structure cleanup that happened on the master branch in the meantime, and some "personal taste" things :)

I'm not sure yet whether I'll add the emscripten code too before merging back to master or doing that separately.

**Comment by floooh** [comment]
Just found a long-standing bug, _sapp_win32_wide_to_utf8() wouldn't zero terminate the result string if it doesn't fit into the buffer. 

This was only used in the clipboard code before.

I'm fixing this now so that no conversion takes place if the string wouldn't fit the buffer (because simply putting a zero byte into the last buffer position might result in a corrupted UTF-8 string if this is in the middle of a multi-byte sequence).

**Comment by prime31** [comment]
> Just found a long-standing bug, _sapp_win32_wide_to_utf8() wouldn't zero terminate the result string if it doesn't fit into the buffer.

Good find! I’m not sure how I missed that when looking through that method.

**Comment by floooh** [comment]
> Good find! I’m not sure how I missed that when looking through that method.

TBH from the documentation of WideCharToMultiByte() I was assuming that the function would zero-terminate the string if it needs to be clamped. But it was easy to test by setting a very small max_dropped_file_path_length.

**Comment by floooh** [comment]
FYI I've done some minor public API changes in my merge-branch:

- renamed the event to ```SAPP_EVENTTYPE_FILES_DROPPED``` (file => files)
- add a ```bool enable_dragndrop``` item to sapp_desc, similar to ```enable_clipboard```, by default this is false, and the whole drag'n'drop system is deactivated (e.g. no allocations taking place)
- bump the default path length to 2048 bytes (a few more KBytes doesn't matter much, and this should be a pretty safe max path length on all systems)

I'm also adding demo code to the events-sapp sample:

image

image

The mouse deltas don't make much sense, because at least on windows, no regular mouse movement is reported during a drag'n'drop operation, but I guess that's fine (the only important info is the mouse position)

**Comment by floooh** [comment]
Ok, I have also added Linux drag'n'drop (taken from GLFW). Some more testing and updating the documentation and then I'll merge this PR back to master.

I haven't implemented anything for HTML5 yet, I'll do that in a separate branch.

**Comment by floooh** [comment]
Ok, this has been merged.

Initial documentation (no WASM/HTML5 yet):

https://github.com/floooh/sokol/blob/03dec6b442ba8111324b879bd3ed4c3241109265/sokol_app.h#L502-L577

**Comment by prime31** [comment]
Woohoo! Can finally get off the fork and back to the mainline. Thanks @floooh!

**Pull request merged**